controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/AbstractRequestProcessor.java

```diff
@@ -0,0 +1,184 @@
+                                    if (ex != null && writeBackPredicate.test(ex)) {
+                                        suppressException(streamMetadataStore.createWaitingRequestIfAbsent(scope, stream, getProcessorName(), context, executor),
+                                                null, "Exception while trying to create waiting request. Logged and ignored.")
+                                                .thenCompose(ignore ->  retryIndefinitelyThenComplete(() -> task.writeBack(event), resultFuture, ex));
+                                    } else {
+                                        // Processing was done for this event, whether it succeeded or failed, we should remove
+                                        // the waiting request if it matches the current processor.
+                                        // If we dont delete it then some other processor will never be able to do the work.
+                                        // So we need to retry indefinitely until deleted.
+                                        retryIndefinitelyThenComplete(() -> streamMetadataStore.deleteWaitingRequestConditionally(scope,
+                                                stream, getProcessorName(), context, executor), resultFuture, ex);
+                                    }
+                                });
+                    } else {
+                        log.debug("Found another processing requested by a different processor {}. Will postpone the event {}.", waitingRequestProcessor, event);
+                        // This is done to guarantee fairness. If another processor has requested for processing
+                        // on this stream, we will back off and postpone the work for later.
+                        retryIndefinitelyThenComplete(() -> task.writeBack(event), resultFuture,
+                                StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED, "Postponed "
+                                        + event + " so that waiting processor" + waitingRequestProcessor + " can work. "));
+                    }
+                });
+
+        return resultFuture;
+    }
+
+    private <R> CompletableFuture<R> suppressException(CompletableFuture<R> future, R returnOnException, String message) {
+        return future.exceptionally(e -> {
+            log.debug(message, e);
+            return returnOnException;
+        });
+    }
+
+    private CompletableFuture<Void> retryIndefinitelyThenComplete(Supplier<CompletableFuture<Void>> futureSupplier, CompletableFuture<Void> toComplete,
+                                                                  Throwable e) {
+        String failureMessage = String.format("Error writing event back into stream from processor %s", getProcessorName());
+        return Retry.indefinitelyWithExpBackoff(failureMessage)
+                .runAsync(futureSupplier, executor)
+                .thenRun(() -> {
```

Comment:
I assume that `runAsync` can NEVER throw an exception, can it? If a single exception sneaks through, this will not execute and your future will be left hanging forever.

I would sleep much better at night if we changed this to a `whenComplete` and always check for an exception there too. You can use `addSuppressed` in that case to record both exceptions.